### ISSUE TITLE
Merge old changes from khval

### ISSUE THREAD

**Issue by samo79** [koobo/HippoPlayer#13]
There were some old changes implemented by khval in late 2017, beside other things some may solve certain issue at exit when running HippoPlayer under AmigaOS4

https://github.com/khval/HippoPlayer

Would be nice if you can merge that code (or atleast a usefull part) into the latest uptodate repository

**Comment by koobo** [comment]
I think the exit problem was that DOS library was used after it was closed, this was fixed in the main repo some time ago.

**Comment by koobo** [comment]
Does it work on AOS4 nowadays?

**Comment by khval** [comment]
Yes, last time I tried with AHI player, it did but the scope did not work.
One of the changed I did, was replace the blit code to blank the scope.

**Comment by khval** [comment]
Maybe using identify.libray can be a idea in order to disable players that’s requires hardware banging.

**Comment by koobo** [comment]
Originally the scope stuff was disabled when AHI is used.

**Comment by khval** [comment]
Yes, I was trying or hopping to understand normal Paula play routines, to the point I can convert this to AHI, but I did get that far. I really suck at 680x0 assembler.

**Comment by khval** [comment]
The only thing I was really happy about was changing the build system, so I did not need to use UAE :-)

**Comment by khval** [comment]
I also did a lot of translation of the comments, in order try to understand the code.

**Comment by koobo** [comment]
A good effort for sure, it  is a difficult challenge!
In the post-v2.45 sources the new changes have comments in English, also some of the old comments have been translated. Additionally there is tons of debug log available.

**Comment by khval** [comment]
Yes, last time I tried with AHI player, it did but the scope did not work.

ons. 15. jun. 2022 kl. 22:26 skrev Kari-Pekka Koljonen <
***@***.***>:

**Comment by koobo** [comment]
In v2.56 there is a limited AHI scope support as well now. Also the scopes are able to work without blitter (if the screen buffer is detected to be non-chip memory). No idea if this actually works on AOS :-)

**Comment by khval** [comment]
Thanks, yes the scopes now work in AHI mode, but non-chip memory detection does not work, OS4.1 does report chip memory (just to make programs happy), it will report 50mb or something, some Vampire system might also report more chip ram then normal but believe it’s a bit less 8mb or something, so system with more then 16mb, you can assume does not have a blitter.

**Comment by khval** [comment]
It’s clearing of the scope that is not working.

**Comment by koobo** [comment]
I think a fool proof way for detecting blitter would be good, or skip direct blitter access and use BltBitmap, which probably would probably work everywhere (but slower).

**Comment by khval** [comment]
Yes OS method will work, I have noticed the scopes will takes a bit of CPU power, it’s not a major problem, but I think it might be running on higher refresh rate then it should.

**Comment by koobo** [comment]
Random thought, does AOS4 and PPC-amigas have working VBlank, that is, does graphics.library/WaitTOF() actually wait a while or is it just a no-operation?

**Comment by khval** [comment]
Yes, of cause it does, but it does not care if its NTSC or PAL, It's 60 FPS VBL or the same as the P96 VBL (at least on AmigaONE's).
DOS / Delay() is 1/50 sec, it has something do with amiga ticks I believe.

**Comment by khval** [comment]
If the scopes takes more than 20ms to render on a Classic Amiga (Amiga500/1200), 
you get 25fps, a more powerful computer it can complete in under 20ms, its able draw it every 50 fps.
(16.6 ms per frame at 60 hz)

let’s say it completes it under 16.6ms (emulation overhead), with a 0.6ms wait, VBL 
(99% of the time is used for rendering)

vs lets say its completed it in 21ms on Classic Amiga, and you mis the first VBL, so wait every 2en VBL,
then you have 40ms at 25fps, that means only ~50% of time is used for render.

**Comment by koobo** [comment]
All the original scopes were designed to run at 50 FPS on the A500 using WaitTOF() sync. 
The reason the VBlank thing came up was that the WInUAE based AmiKit Amiga desktop had  problems as it had "Hardware VBlank emulation" turned off. This caused weird things happening, for example WaitTOF() did nothing. The scopes worked but basically ran at full speed, eating all CPU cycles.
